### PR TITLE
Sumdigits exmaple with GRPO

### PR DESCRIPTION
**Pull request by Ritesh1905**
The reward-mean converges in less than 15 training steps and takes ~5 mins 

wandb run: https://meta.wandb.io/torchforge/sumdigits-training/runs/uxzowpkp?nw=nwuserrithesh

image

- Made some corrections to GRPO loss function.
- Updates the sumdigits example to use GRPO loss
- Introduces a small curriculum setup.

TODO:

- Lot of code redundancy, which will be fixed once the abstractions are landed.

### CODE REVIEW

File: src/forge/util/model_utils.py
Comment (joecummings) [inline src/forge/util/model_utils.py:1]: nit: util/model_utils.py -> utils/ops.py

File: src/forge/util/model_utils.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import torch.nn.functional as F
+
+
+def selective_log_softmax(logits, index) -> torch.Tensor:
```
Comment (joecummings) [inline src/forge/util/model_utils.py]: Could be include a quick test for this as well? Plus some typing for logits and index

File: src/forge/losses/grpo_loss.py [hunk 2]
```diff
@@ -9,21 +9,70 @@
 
 
 class SimpleGRPOLoss(nn.Module):
-    """Simplified GRPO Loss for simplified single step updates
-    Inspired by the Hugging Face TRL implementation:
-        https://github.com/huggingface/trl/blob/417915a3e4d3e3bc8d7b196594308b8eabf928be/trl/trainer/grpo_trainer.py#L1624.
+    """
+    Group Relative Policy Optimization (GRPO) Loss Function.
```
Comment (joecummings) [inline src/forge/losses/grpo_loss.py]: Let's discuss offline. In general, I think we should have a GRPOLoss that follows exactly the original spec: https://huggingface.co/docs/trl/en/grpo_trainer#loss-types and then something like DrGRPO which is shown to be more stable wrt length and difficulty bias: https://huggingface.co/docs/trl/en/grpo_trainer#loss-types

File: src/forge/losses/grpo_loss.py [hunk 3]
```diff
@@ -9,21 +9,70 @@
...
+        """
+        # Compute log ratio: log(π_θ/π_ref) = log(π_θ) - log(π_ref)
+        log_ratio = logprobs - ref_logprobs
+
+        # Policy gradient loss: -log_ratio * advantages
+        per_token_policy_loss = log_ratio * advantages
+
+        # KL penalty: KL(π_θ || π_ref) ≈ π_θ * (log(π_θ) - log(π_ref))
+        # For log probabilities, this becomes: exp(logprobs) * (logprobs - ref_logprobs)
+        kl_penalty = torch.exp(logprobs) * (logprobs - ref_logprobs)
```
Comment (Ritesh1905) [inline src/forge/losses/grpo_loss.py]: @joecummings , my bad, This is incorrect... It should be  same as original implementation.

```
log_ratio = ref_logprobs - logprobs
kl_penalty = torch.exp(log_ratio) - log_ratio - 1
```

I will revert the chnages to this file

Comment (Ritesh1905) [reply src/forge/losses/grpo_loss.py]: reverted changes to this file.
